src/snowflake/connector/auth/okta.py

```diff
@@ -266,26 +271,37 @@ def _step3(
     @staticmethod
     def _step4(
         conn: SnowflakeConnection,
-        one_time_token: str,
+        generate_one_time_token: Callable,
         sso_url: str,
     ):
         logger.debug("step 4: query IDP URL snowflake app to get SAML " "response")
-        url_parameters = {
-            "RelayState": "/some/deep/link",
-            "onetimetoken": one_time_token,
-        }
-        sso_url = sso_url + "?" + urlencode(url_parameters)
-        headers = {
-            HTTP_HEADER_ACCEPT: "*/*",
-        }
-        response_html = conn._rest.fetch(
-            "get",
-            sso_url,
-            headers,
-            timeout=conn.login_timeout,
-            socket_timeout=conn.login_timeout,
-            is_raw_text=True,
-        )
+        one_time_token = generate_one_time_token()
+        timeout_time = time.time() + conn.login_timeout if conn.login_timeout else None
+        response_html = {}
+        while timeout_time is None or time.time() < timeout_time:
+            try:
+                url_parameters = {
+                    "RelayState": "/some/deep/link",
+                    "onetimetoken": one_time_token,
+                }
+                sso_url = sso_url + "?" + urlencode(url_parameters)
+                headers = {
+                    HTTP_HEADER_ACCEPT: "*/*",
+                }
+                remaining_timeout = timeout_time - time.time() if timeout_time else None
+                response_html = conn._rest.fetch(
+                    "get",
+                    sso_url,
+                    headers,
+                    timeout=remaining_timeout,
+                    socket_timeout=remaining_timeout,
+                    is_raw_text=True,
+                    is_okta_authentication=True,
+                )
+                break
+            except RefreshToken:
+                logger.debug("step4: refresh token for re-authentication")
+                one_time_token = generate_one_time_token()
```

Comment:
This could move to beginning of loop (make sure to remove one call from before the loop)